buildSrc/subprojects/parser/src/test/java/org/apache/calcite/test/DialectGenerateTest.java

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.test;
+
+import java.nio.file.Path;
+import java.io.File;
+import java.nio.file.Paths;
+import java.util.Queue;
+import java.util.LinkedList;
+import java.lang.StringBuilder;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.LinkedHashMap;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Assertions;
+import org.apache.calcite.buildtools.parser.DialectGenerate;
+import org.apache.calcite.buildtools.parser.DialectGenerate.CurlyParser;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DialectGenerateTest {
+
+  /**
+   * Retrns a DialectGenerate with root path of calcite/parsingTest and
+   * dialect path of calcite/parsingTest/intermediate/testDialect.
+   */
+  private DialectGenerate setupDialectGenerate() {
+    Path rootPath = Paths.get("..", "..", "..", "parsingTest");
+    // Adds the path /intermediate/testDialect/ to the end of rootPath.
+    Path dialectPath = rootPath.resolve(Paths.get("intermediate", "testDialect"));
+    File rootFile = rootPath.toFile();
+    File dialectFile = dialectPath.toFile();
+    return new DialectGenerate(dialectFile, rootFile, "");
+  }
+
+  /**
+   * Returns the expected extraction results after calling DialectGenerate.extractFunctions
+   * onto the parsingTest directory structure.
+   */
+  private Map<String, String> getExpectedExtractionResults() {
+    String foo = "void foo() :\n"
+      + "{\n"
+      + "    String x = \" '}' \";\n"
+      + "}\n"
+      + "{\n"
+      + "    // overidden by testDialect\n"
+      + "}";
+    String bar = "void bar  (   )  : {} {}";
+    String baz = "void baz() : {x}\n"
+      + "{\n"
+      + "    // overridden by intermediate\n"
+      + "}";
+    String qux = "void qux(int arg1, int arg2) :\n"
+      + "{\n"
+      + "    String x = \" } \"\n"
+      + "}\n"
+      + "{\n"
+      + "    // Below is a }\n"
+      + "}";
+    String quux = "void quux(int arg1,\n"
+      + "    int arg2\n"
+      + ") :\n"
+      + "{\n"
+      + "    char x = ' } '\n"
+      + "    String y;\n"
+      + "}\n"
+      + "{\n"
+      + "    /* All invalid curly braces:\n"
+      + "    }\n"
+      + "    // }\n"
+      + "    \" } \"\n"
+      + "    ' } '\n"
+      + "    */\n"
+      + "    <TOKEN> {\n"
+      + "        y = \" { } } \"\n"
+      + "    }\n\n"
+      + "    // Not a string: \"\n"
+      + "}";
+    return new LinkedHashMap<String, String>() {{
+      put("foo", foo);
+      put("bar", bar);
+      put("baz", baz);
+      put("qux", qux);
+      put("quux", quux);
+    }};
+  }
+
+  private Queue<String> getTokens(String str) {
+    return new LinkedList(Arrays.asList(DialectGenerate.tokenizerPattern.split(str)));
```

Comment:
nit: Line wrap

Reply:
nit: Also, consider keeping tokenizerPattern private and instead providing a public getTokens() function in DialectGenerate.